src/compile/projection/parse.ts

```diff
@@ -0,0 +1,65 @@
+import {SHAPE, X, Y} from '../../channel';
+import {Config} from '../../config';
+import {MAIN} from '../../data';
+import {isProjectionFieldDef} from '../../fielddef';
+import {GEOSHAPE} from '../../mark';
+import {Projection} from '../../projection';
+import {contains} from '../../util';
+import {VgProjection} from '../../vega.schema';
+import {Model} from '../model';
+import {UnitModel} from '../unit';
+import {ProjectionComponent} from './component';
+
+export function parseProjection(model: Model, config: Config) {
+  if (model instanceof UnitModel) {
+    model.component.projection = parseUnitProjection(model, config);
+  } else {
+    const mergedProjection = parseNonUnitProjections(model);
+  }
+}
+
+function parseUnitProjection(model: UnitModel, config: Config): ProjectionComponent {
+  const {projection, markDef, encoding} = model;
+
+    // TODO: make sure this actually works for both kinds of lookups, with GeoJSON transform, and with basic projection
+    const data = model.requestDataName(MAIN);
+
+    const width = model.getSizeSignalRef('width');
+    const height = model.getSizeSignalRef('height');
+    const size = [width.signal, height.signal];
+
+    const isGeoshapeMark = markDef && markDef.type === GEOSHAPE;
+    const hasEncodedProjection = encoding && [X, Y, SHAPE].some((channel) => isProjectionFieldDef(encoding[channel]));
+    const inheritedProjection = config && config.projection;
+
+    if (model.projection || (inheritedProjection && (isGeoshapeMark || hasEncodedProjection))) {
+      return new ProjectionComponent(model.getName('projection'), {
+        ...config && config.projection ? config.projection : {},
+        ...(projection || {}),
+      } as Projection, size, data);
+  }
+
+  return undefined;
+}
+
+function parseNonUnitProjections(model: Model): ProjectionComponent {
+  const merged = model.children.reduce((projections, child) => {
+    const projection = child.component.projection;
+    if (projection) {
+      projections.forEach((other) => {
```

Comment:
Here you seems to always merge -- see `parseNonUnitScaleCore` how I make two passes over the children.  The first pass checks whether I can merge all of them.  If can merge, then make second pass to really merge.